genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GrpcHeartBeatServiceImpl.java

```diff
@@ -0,0 +1,244 @@
+
+    private void handleAgentHeartBeat(
+        final String streamId,
+        final AgentHeartBeat agentHeartBeat
+    ) {
+        // Pull the record, if one exists
+        final AgentStreamRecord agentStreamRecord;
+        synchronized (activeStreamsMap) {
+            agentStreamRecord = activeStreamsMap.get(streamId);
+        }
+
+        final String claimedJobId = agentHeartBeat.getClaimedJobId();
+        if (agentStreamRecord == null) {
+            log.warn("Received heartbeat from an unknown stream");
+        } else if (StringUtils.isBlank(claimedJobId)) {
+            log.warn("Ignoring heartbeat lacking job id");
+        } else {
+            log.info("Received heartbeat from agent that claimed job: {}", claimedJobId);
+            final boolean isFirstHeartBeat = agentStreamRecord.updateRecord(claimedJobId);
+            // On first heartbeat, notify listeners of a new agent connection
+            if (isFirstHeartBeat) {
+                notifyAgentConnected(agentStreamRecord.getJobId());
+            }
+        }
+    }
+
+    private void handleStreamCompletion(final String streamId) {
+        // Pull the record, if one exists
+        final AgentStreamRecord agentStreamRecord;
+        synchronized (activeStreamsMap) {
+            agentStreamRecord = activeStreamsMap.remove(streamId);
+        }
+
+        if (agentStreamRecord == null) {
+            log.warn("Received completion from an unknown stream");
+        } else {
+            agentStreamRecord.responseObserver.onCompleted();
+            if (agentStreamRecord.hasJobId()) {
+                notifyAgentDisconnected(agentStreamRecord.getJobId());
+            }
+        }
+    }
+
+    private void handleStreamError(final String streamId, final Throwable t) {
+        // Pull the record, if one exists
+        final AgentStreamRecord agentStreamRecord;
+        synchronized (activeStreamsMap) {
+            agentStreamRecord = activeStreamsMap.remove(streamId);
+        }
+
+        if (agentStreamRecord == null) {
+            log.warn("Received error from an unknown stream");
+        } else {
+            agentStreamRecord.responseObserver.onError(t);
+            if (agentStreamRecord.hasJobId()) {
+                notifyAgentDisconnected(agentStreamRecord.getJobId());
+            }
+        }
+    }
+
+    private void notifyAgentConnected(final String jobId) {
+        agentRoutingService.handleClientConnected(jobId);
+    }
+
+    private void notifyAgentDisconnected(final String jobId) {
+        agentRoutingService.handleClientDisconnected(jobId);
+    }
+
+    private static class AgentStreamRecord {
+        private final StreamObserver<ServerHeartBeat> responseObserver;
+        private String claimedJobId;
+
+        AgentStreamRecord(
+            final StreamObserver<ServerHeartBeat> responseObserver
+        ) {
+            this.responseObserver = responseObserver;
+        }
+
+        synchronized boolean updateRecord(final String jobId) {
```

Comment:
@mprimi For my understanding, how will claimedJobId be different than jobId.

Reply:
"claimed job id" is a job id that we confirmed with the server was claimed by the agent for execution.